lib/srv/db/redis/engine.go

```diff
@@ -0,0 +1,468 @@
+)
+
+// Engine implements common.Engine.
+type Engine struct {
+	// Auth handles database access authentication.
+	Auth common.Auth
+	// Audit emits database access audit events.
+	Audit common.Audit
+	// Context is the database server close context.
+	Context context.Context
+	// Clock is the clock interface.
+	Clock clockwork.Clock
+	// Log is used for logging.
+	Log logrus.FieldLogger
+	// proxyConn is a client connection.
+	proxyConn net.Conn
+	// clientReader is a go-redis wrapper for Redis client connection.
+	clientReader *redis.Reader
+	// sessionCtx is current session context.
+	sessionCtx *common.Session
+}
+
+// InitializeConnection initializes the database connection.
+func (e *Engine) InitializeConnection(clientConn net.Conn, sessionCtx *common.Session) error {
+	e.proxyConn = clientConn
+	e.clientReader = redis.NewReader(clientConn)
+	e.sessionCtx = sessionCtx
+
+	return nil
+}
+
+// authorizeConnection does authorization check for MongoDB connection about
+// to be established.
+func (e *Engine) authorizeConnection(ctx context.Context, sessionCtx *common.Session) error {
+	ap, err := e.Auth.GetAuthPreference(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	mfaParams := services.AccessMFAParams{
+		Verified:       sessionCtx.Identity.MFAVerified != "",
+		AlwaysRequired: ap.GetRequireSessionMFA(),
+	}
+
+	dbRoleMatchers := role.DatabaseRoleMatchers(
+		sessionCtx.Database.GetProtocol(),
+		sessionCtx.DatabaseUser,
+		// Leave database empty as Redis integration doesn't support schema access control.
+		"",
+	)
+	err = sessionCtx.Checker.CheckAccess(
+		sessionCtx.Database,
+		mfaParams,
+		dbRoleMatchers...,
+	)
+	if err != nil {
+		e.Audit.OnSessionStart(e.Context, sessionCtx, err)
+		return trace.Wrap(err)
+	}
+	return nil
+}
+
+// SendError sends error message to connected client.
+func (e *Engine) SendError(redisErr error) {
+	if redisErr == nil || utils.IsOKNetworkError(redisErr) {
+		return
+	}
+
+	e.Log.Debugf("sending error to Redis client: %v", redisErr)
+
+	if err := e.sendToClient(redisErr); err != nil {
+		e.Log.Errorf("Failed to send message to the client: %v", err)
+		return
+	}
+}
+
+// sendToClient sends a command to connected Redis client.
+func (e *Engine) sendToClient(vals interface{}) error {
+	buf := &bytes.Buffer{}
+	wr := redis.NewWriter(buf)
+
+	if err := writeCmd(wr, vals); err != nil {
+		return trace.BadParameter("Failed to convert error to a message: %v", err)
+	}
+
+	if _, err := e.proxyConn.Write(buf.Bytes()); err != nil {
+		return trace.ConnectionProblem(err, "Failed to send message to the client")
+	}
+
+	return nil
+}
+
+// HandleConnection is responsible for connecting to a Redis instance/cluster and
+func (e *Engine) HandleConnection(ctx context.Context, sessionCtx *common.Session) error {
+	// Check that the user has access to the database.
+	err := e.authorizeConnection(ctx, sessionCtx)
+	if err != nil {
+		return trace.Wrap(err, "error authorized database access")
```

Comment:
Nit: I think user message is redundant here.
```suggestion
		return trace.Wrap(err)
```